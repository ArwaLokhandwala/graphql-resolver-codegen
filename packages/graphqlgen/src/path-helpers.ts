import * as fs from 'fs';
import * as path from 'path'

// TODO write test cases
export function getAbsoluteFilePath(modelPath: string): string {
  let absolutePath = path.resolve(modelPath)

  if (!fs.existsSync(absolutePath) && fs.existsSync(`${absolutePath}.ts`)) {
    absolutePath += '.ts'
  }

  if (!fs.lstatSync(absolutePath).isDirectory()) {
    if (path.extname(absolutePath) !== '.ts') {
      throw new Error(`${absolutePath} has to be a .ts file`)
    }

    return absolutePath
  }

  if (!fs.existsSync(absolutePath)) {
    throw new Error(`${absolutePath} not found`)
  }

  const indexTsPath = path.join(absolutePath, 'index.ts')
  if (!fs.existsSync(indexTsPath)) {
    throw new Error(`No index.ts file found in directory: ${absolutePath}`)
  }

  return indexTsPath
}

// TODO write test cases
export function getImportPathRelativeToOutput(
  absolutePath: string,
  outputDir: string,
): string {
  let relativePath = path.relative(path.dirname(outputDir), absolutePath)

  if (!relativePath.startsWith('.')) {
    relativePath = './' + relativePath
  }

  // remove .ts file extension
  relativePath = relativePath.replace(/\.ts$/, '')

  // remove /index
  relativePath = relativePath.replace(/\/index$/, '')

  return relativePath
}