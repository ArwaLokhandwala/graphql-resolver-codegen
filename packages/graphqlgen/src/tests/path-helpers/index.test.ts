import { getAbsoluteFilePath } from '../../path-helpers'
import { join } from 'path'

test('invalid path', () => {
  expect(() => getAbsoluteFilePath(join('../fixtures/type.ts'))).toThrow()
})

test('valid path', () => {
  expect(
    getAbsoluteFilePath(join(__dirname, '../fixtures/basic/types.ts'))
      .split('/')
      .slice(-4),
  ).toMatchSnapshot()
})
