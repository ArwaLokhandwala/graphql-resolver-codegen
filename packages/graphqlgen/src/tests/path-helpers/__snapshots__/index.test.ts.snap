// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valid path 1`] = `
Array [
  "tests",
  "fixtures",
  "basic",
  "types.ts",
]
`;
